### PR TITLE
fix(hooks): deliver run-start hook context in the SDK runtime; core-layer blocking is opt-in

### PR DESCRIPTION
**Pull request by mkondratek**
## Problem

UserPromptSubmit/TaskStart/TaskResume hooks can't inject context on the next engine — the remaining slice of the regression #13297/#13298 fixed for tool hooks. Worse, at the SDK core layers the run-start events (`agent_start`/`agent_resume`) spawned detached with stdio ignored, so their entire output was discarded: neither `contextModification` nor `cancel` was honored. Legacy delivered these as `<hook_context>` blocks in the same turn as the user prompt. Note the core-layer discard is *kept as the default* here and the capability put behind a flag — see the SDK-core bullet below for why.

Ticket: [CLINE-3058](https://linear.app/cline-bot/issue/CLINE-3058/non-tool-hook-events-userpromptsubmittaskstarttaskresume-cant-inject) (split out of [CLINE-2987](https://linear.app/cline-bot/issue/CLINE-2987), source report #13277)

**Scope**: this PR is the SDK half (`sdk/`, the CLI's message-added filter, hooks README). The VS Code adapter mapping and the live e2e that proves injection end to end are stacked on top in #14160.

## Change

Mirrors the tool-hook `appendContext` shape one level up. All of it is provider-agnostic runtime/contract plumbing — no provider-specific code.

- **Contract** (`sdk/packages/shared/src/agent.ts`): `beforeRun` returns a new `AgentRunStartResult` — `stop`/`reason` plus an `appendContext` channel. `AgentStopControl` remains assignable, so existing implementors compile unchanged.
- **Runtime** (`sdk/packages/agents/src/agent-runtime.ts`): run-start context is collected in `callBeforeRunHooks` and injected as one `<hook_context source="RunStart">` user message (hidden display role, like the tool variant) **after the run's input messages**, so the model sees it on the run's first request — matching where legacy placed it. One formatter builds every hook-context block, keyed by a discriminated origin (run-start vs tool hook), so trimming, tag neutralizing and attribute escaping live in one place. The post-tool flush is shared with the run-start flush, the buffer is cleared at run start, and a stopping hook's context is never injected. `mergeHookFunction` already concatenates `appendContext` generically, so multi-layer merging needed no changes.
- **Resume edge**: when a resumed session is seeded with a trailing unresolved `tool_use`, the run-start block goes in immediately *before* that call rather than after it, keeping the `tool_use`/`tool_result` pair adjacent while still delivering the context. (An earlier revision deferred the flush instead, which only delivered if the model happened to call a tool next — per review.)
- **SDK core layers** (`hook-file-hooks.ts`, `subprocess.ts`): blocking run-start dispatch is **opt-in** via a new `blockingRunStartHooks` option, defaulting to today's fire-and-forget. These hooks have always run detached with stdio ignored, so blocking by default would silently stall every run start for hosts whose existing scripts linger (per review). With the flag on, `agent_start`/`agent_resume` run blocking with their control honored — `cancel` stops the run with the hook's message as the reason, context becomes `appendContext` with the existing 50KB truncation — and long-running run-start hooks must background themselves (spawn a detached child and exit, which the settle-after-exit change explicitly supports). `agent_resume` shares the path, completing TaskResume at these layers once the flag flips.
- **`hooks.detached_runtime` telemetry**: the detached path reports each hook exactly once — on exit with its runtime, or, if the hook is still running after a 30s observation window, a censored sample (`exited: false`, duration as a lower bound). The censored half is the point: hooks that linger are the ones that decide whether blocking is safe, and an exit-only observer would have recorded only the harmless population. A hook that fails to spawn produces no sample. Nothing is awaited and the timer is unref'd, so latency is unchanged. Today no surface can answer "how long do run-start hooks take?" (the ones reporting durations are the ones already running hooks blocking), which is exactly why the flag defaults off; this makes the flip a measurement rather than a guess.
- Drive-by: the CLI's and the audit logger's `prompt_submit` message-added dispatch no longer fires for injected hook-context messages (user role, system display role), which they previously mistook for user prompts.

## Not in this PR

- **The VS Code adapter and e2e** — stacked in #14160.
- **CLI/desktop don't flip `blockingRunStartHooks` yet** — run-start control (cancel/context) stays inert on those surfaces until a deliberate flip with a changelog note for hook authors; CLINE-3058 remains open for them.
- **Core-layer `prompt_submit` stays fire-and-forget**: when `beforeRun` runs at those layers the run's input isn't in the snapshot yet, so message-added is the only point with the prompt text — and it has no return channel. Documented in code. (Correcting an earlier figure here: CLI hook usage is *not* negligible — `hooks.discovery_completed` shows ~1,015 CLI users with `prompt_submit` scripts and ~1,064 with `agent_start` over 30 days. My earlier "~2 users" came from querying the near-empty `CLI` surface instead of `cli`. This makes the opt-in default more clearly right, and makes the eventual flip worth doing deliberately.)

## Review rounds

Fixed along the way, mostly from self-review and review comments:

- **The big one:** the session-runtime orchestrator's `beforeRun` merger returned `undefined` for every non-stopping result, discarding `appendContext` on every production path (VS Code, CLI, desktop) — only directly-constructed runtimes saw the feature. It now aggregates like its `beforeModel`/`beforeTool` siblings, with orchestrator-level tests.
- Injected hook-context messages no longer masquerade as user prompts anywhere: the `message-added` dispatchers (CLI, audit logger) filter on the system display role.
- A blocking hook that spawns a background child sharing its stdout could hang the run forever (completion resolved on stdio close, which the grandchild holds past the SIGKILL). The runner now settles after a named post-exit grace period and drops the hook's stdio pipes at that point, so the lingering grandchild neither keeps the host alive nor keeps filling the output buffers (hook authors: give a backgrounded child its own stdio); covered by tests.
- Run-start hooks in the subprocess layer tolerate non-JSON stdout (their output was ignored for their whole prior life) instead of erroring every run; timeouts still surface.
- A stopping layer short-circuits the layer merge, duplicate control mappers collapsed, hooks README updated with the blocking contract and per-host timeouts.

**Open questions for review**: (1) ~~run-start timeout~~ — superseded by the opt-in gate: the core layers default to fire-and-forget, and the 120s budget only applies once a host deliberately opts into blocking; a tighter default and abort-signal threading can ride along with that flip. (2) The generic `source="RunStart"` label loses TaskStart-vs-UserPromptSubmit attribution and shows as a "RunStart" chip on transcript replay — per-hook attribution needs a structured `appendContext` contract, proposed as a follow-up. (3) On orchestrated sessions the prompt *is* in the snapshot at `beforeRun` (seeded via `initialMessages`, never emitted as `message-added`) — meaning core-layer `prompt_submit` could move to `beforeRun` and gain injection; left as a follow-up candidate.

## Testing

- **Live e2e** proving injection end to end in a real VS Code lives in #14160; it passed on the combined branch before the split.
- Runtime tests: `beforeRun` `appendContext` lands as a `RunStart` hook-context message after the prompt on the run's first model request (hidden display role asserted); with a seeded trailing tool call the block sits immediately before that call and the call stays last in the request; a stopping hook's context is never injected; embedded `hook_context` tags and tool attributes keep their coverage through the unified formatter. `@cline/agents`: 73 pass.
- Core tests: a `TaskStart.js` hook file's `contextModification` comes back as `appendContext` from `beforeRun` and `cancel` returns `stop` (both with the flag on); with the flag off the same hook's control is ignored; a detached hook's runtime is reported through `onHookRuntime`; the orchestrator merger aggregates `appendContext` across config and extension layers; a hook whose child holds stdout open still settles and the host releases its pipes; a detached hook that fails to spawn produces no runtime sample. Full core unit suite: 2402 pass, 108 skipped, plus one pre-existing flaky bash-executor test that fails under the full run and passes in isolation and on clean `main` (untouched here).
- `apps/cli` hooks test: 7 pass. Shared/agents/core typechecks clean; `apps/vscode` `tsc --noEmit` clean against these SDK changes with `main`'s adapter.

### CODE REVIEW

**Review by abeatrix — CHANGES_REQUESTED** [review]
Nice work tracking down the full regression slice here — the orchestrator-merger catch in the self-review round and the e2e that proves injection end-to-end are exactly what this needed. The runtime/contract plumbing (AgentRunStartResult, the flush placement, the pairing-rule deferral) and the adapter changes all look right to me, and they change nothing for hosts that don't opt in.

One design concern before this merges, on the SDK-core hunks (hook-file-hooks.ts / subprocess.ts): making agent_start/agent_resume blocking by default silently changes run-start latency semantics for every existing CLI/desktop user with a run-start hook file. Those hooks have been fire-and-forget with stdio ignored for their entire life, so a daemon-style script that lingers (previously harmless) now stalls every run start for the full 120s timeout — and the 1s-after-exit settle doesn't help a script that never exits itself. Your own data point suggests the population that wants the new contract on this path is tiny (~2 prompt_submit users/30d), while everyone with an existing run-start hook is exposed to the regression.

**Proposal**: keep everything else as-is, and gate the core-layer blocking behind an option (e.g. blockingRunStartHooks on the hook-extension/subprocess options), defaulting to the current fire-and-forget. The CLI/desktop can flip it deliberately — with a changelog note telling hook authors that long-running run-start hooks must background themselves (spawn detached + exit, which the new settle-after-exit explicitly supports). Trade-off stated honestly: this leaves CLINE-3058 open for CLI/desktop until the flag flips, and forks run-start semantics between hosts in the interim — I think that's the right side of the trade given the silent-stall risk, but flagging it as a judgment call, not a blocker.

If you'd rather ship blocking-by-default, then two asks instead: (1) resolve your open question #1 toward the extension's 30s rather than inheriting the 120s tool-hook budget — the same hook file shouldn't be able to stall a CLI run 4× longer than an extension run; (2) a release-note callout for the backgrounding pattern above.

**Comment by mkondratek** [comment]
thanks Bee. your idea makes sense. applied.

I did a query too, and the truth is we can't tell today how long those hooks run. The surfaces that report hook durations are the ones that already run hooks blocking, so for CLI we have nothing — and ~1k users do have `agent_start` scripts (my earlier "~2 users" was a bad query on my side).

I piggybacked the telemetry on it: the detached hook will report how long it runs and whether it exited at all. Nothing is awaited, so latency stays the same. In a month we should know if we can flip the default.

**Review by abeatrix — CHANGES_REQUESTED** [review]
The main hook-context fix looks correct and well tested, but I’m requesting one change before approval.

In `subprocess-runner.ts`, the new detached-hook telemetry claims to report whether a hook exited. However, `onDetachedSettled` is only invoked after `completed` resolves and always reports `exited: true`. A hook that never exits emits no event, so the telemetry is biased toward completed hooks and cannot reliably support a future decision to make run-start hooks blocking by default.

Please either:

- emit a censored `exited: false` observation after a defined threshold, or
- remove the `exited` field and clearly document that this telemetry only measures hooks that completed.

Everything else in the PR looks good to me, including the opt-in blocking behavior, context aggregation and injection, tool-pairing protection, and test coverage.

**Comment by mkondratek** [comment]
good catch. thank you for staying watchful. 

(option A) emiting exit: false after a threshold APPLIED

thanks!

**Review by abeatrix — COMMENTED** [review]
looks like the windows sdk check is failing cause by the new test

**Comment by saoudrizwan** [comment]
@abeatrix do the latest changes address your review, and does this look good to merge once CI finishes? If so, could you approve it to clear the outstanding changes-requested review? If anything is still missing, please let us know.

File: sdk/packages/core/src/hooks/subprocess-runner.ts [hunk 1]
```diff
@@ -195,7 +222,24 @@ export async function runSubprocessEvent(
...
+		child.once("close", (exitCode) => settle(exitCode));
+		// "close" waits for the stdio pipes to reach EOF, and a hook that spawned
+		// a background child sharing its stdout keeps the pipe open after the
+		// hook itself exits — without a fallback the caller would wait on that
+		// grandchild forever (the timeout's SIGKILL only reaches the direct
+		// child). Settle shortly after process exit with whatever output arrived;
+		// a resolved promise ignores the eventual "close".
+		if (!detached) {
+			child.once("exit", (exitCode) => {
+				const fallback = setTimeout(() => settle(exitCode), 1_000);
```
Comment (abeatrix) [inline sdk/packages/core/src/hooks/subprocess-runner.ts]: [P2] Clean up inherited pipes when the fallback settles. Resolving leaves stdout/stderr and their data listeners active. Reproduced: the hook returned after 1.02s, but Node stayed alive until its background child exited at 3.54s. A persistent child can keep the host alive and keep accumulating output. Add once-only pipe/listener and timer cleanup, with a test that checks host exit after settlement.

File: sdk/packages/core/src/hooks/subprocess-runner.ts [hunk 2]
```diff
@@ -208,6 +252,61 @@ export async function runSubprocessEvent(
...
+			censorTimer.unref?.();
+			void completed
+				.then((result) =>
+					reportOnce({
+						durationMs: Date.now() - startedAt,
+						exitCode: result.exitCode,
+						exited: true,
+					}),
+				)
+				.catch(() => undefined);
```
Comment (abeatrix) [inline sdk/packages/core/src/hooks/subprocess-runner.ts]: [P2] Clear the observation timer when spawning fails. This catch leaves the timer armed, so a nonexistent executable is later reported as exited:false (a false 30-second runtime sample with production defaults). Reproduced with a missing command. Cancel the timer on rejection and omit the runtime sample or report a distinct spawn failure.

**Review by abeatrix — CHANGES_REQUESTED** [review]
The context-injection plumbing looks reasonable, but please fix the two subprocess lifecycle issues below before merging. All 7 subprocess-runner tests passed; additional probes reproduced both issues.

File: sdk/packages/core/src/hooks/subprocess-runner.ts (hunk 1 above)
Comment (mkondratek) [reply sdk/packages/core/src/hooks/subprocess-runner.ts]: good catch, thanks!

The fallback now drops stdin/stdout/stderr (listeners included) right after it settles so the grandchild can't keep the host alive or keep filling the buffers. 

Added a test that counts the process's active pipe handles before and after — it fails on the old code (pipes stay until the grandchild exits) and passes now. Also, added a line to the hooks readme telling authors to give a backgrounded child its own stdio, since anything it writes to the inherited pipes after the hook exits is gone.

File: sdk/packages/core/src/hooks/subprocess-runner.ts (hunk 2 above)
Comment (mkondratek) [reply sdk/packages/core/src/hooks/subprocess-runner.ts]: yep, that was a false sample. spawn failures now cancel the observation timer and emit nothing — no hook ran, so there's no runtime to report, and the failure already surfaces through the rejected run.

test added with a missing executable. it caught it exited: false on the old code

File: sdk/packages/agents/src/agent-runtime.ts [hunk 3]
```diff
@@ -362,6 +362,15 @@ function formatHookContextBlock(
 	return `<hook_context source="${source}" tool_name="${toolName}" tool_call_id="${toolCallId}">\n${body}\n</hook_context>`;
 }
 
+function formatRunHookContextBlock(text: string): string {
+	// Run-start hooks (TaskStart/UserPromptSubmit/TaskResume in their various
+	// layer spellings) have no tool identity; layers merge their outputs before
+	// the runtime sees them, so a single generic source labels the block.
+	// Embedded hook_context tags are neutralized like the tool variant.
+	const body = text.trim().replace(/<(\/?)hook_context/gi, "<\\$1hook_context");
+	return `<hook_context source="RunStart">\n${body}\n</hook_context>`;
```
Comment (abeatrix) [inline sdk/packages/agents/src/agent-runtime.ts]: Maintainability suggestion: could we generalize the existing formatHookContextBlock instead of adding a second formatter with the same text trimming, tag escaping, and wrapper construction? A discriminated union for run-start versus tool-hook context would let one formatter always emit source, and emit tool_name/tool_call_id only for the tool variant. This keeps the escaping rules in one place without passing empty or fabricated tool identities. Please retain coverage for embedded hook_context tags and the existing tool attributes.

File: sdk/packages/agents/src/agent-runtime.ts [hunk 4]
```diff
@@ -937,12 +934,56 @@ export class AgentRuntime {
...
+		// resumed session can be seeded with a trailing unresolved tool call,
+		// and a user message in that gap breaks providers' pairing rules. The
+		// buffer keeps the context until the next flush point, which runs
+		// after the tool messages.
+		const lastMessage = this.state.messages.at(-1);
+		if (
+			lastMessage?.role === "assistant" &&
+			lastMessage.content.some((part) => part.type === "tool-call")
+		) {
+			return;
```
Comment (abeatrix) [inline sdk/packages/agents/src/agent-runtime.ts]: This guard avoids inserting context into an unresolved tool-call/result pair, but deferral does not guarantee delivery. If the next response contains no tool calls, the run can finish without reaching the post-tool flush; the next run clears pendingHookContexts. The new "defers run-start context past a seeded trailing tool call" test actually asserts that resume-context is absent, so it covers dropping the context rather than eventual delivery. Could we either resolve/validate incomplete seeded history before the first request and then inject the context, or explicitly reject this unsupported history shape? Please test the selected behavior instead of silently losing context. This is a narrow resume edge case, not a claim that the normal run-start path is broken; simply removing the pairing guard would not be a safe fix.

**Review by abeatrix — COMMENTED** [review]
Suggestions focused on agent-runtime.ts: consolidate hook-context formatting and make the deferred-context behavior explicit. The shared flush helper and both flush locations are appropriate: run-start context belongs after input and before the first model request, and tool-hook context belongs after tool results. Please also retain the run-start buffer reset so context from an interrupted run cannot leak into a later run.

File: sdk/packages/agents/src/agent-runtime.ts [hunk 5]
```diff
@@ -823,24 +837,7 @@ export class AgentRuntime {
 						message: toolMessage,
 					});
 				}
-				if (this.pendingHookContexts.length > 0) {
-					const hookContextText = this.pendingHookContexts.join("\n\n");
-					this.pendingHookContexts = [];
-					// displayRole "system" keeps the injected block out of user-facing
-					// transcripts (live and replayed) while it still reaches the model,
-					// mirroring how compaction summaries are handled.
```
Comment (abeatrix) [inline sdk/packages/agents/src/agent-runtime.ts:889]: should we keep this comment in the hookContextMessage code inside flushPendingHookContexts function below?

File: sdk/packages/core/src/hooks/subprocess-runner.ts [hunk 6]
```diff
@@ -195,7 +247,30 @@ export async function runSubprocessEvent(
...
+		// child). Settle shortly after process exit with whatever output arrived,
+		// then drop the pipes: resolving alone would leave them (and their data
+		// listeners) attached for as long as the grandchild lives, keeping the
+		// host alive and accumulating output. Destroying them also lets the
+		// child's own "close" fire, which the settled promise then ignores.
+		if (!detached) {
+			child.once("exit", (exitCode) => {
+				exitFallback = setTimeout(() => {
+					settle(exitCode);
+					releaseStdio(child);
```
Comment (abeatrix) [inline sdk/packages/core/src/hooks/subprocess-runner.ts:277]: Nonblocking maintainability suggestion: please extract the 1_000 fallback delay into a named constant and document that it is a grace period for draining output after process exit. Keep releaseStdio paired with fallback settlement: resolving the promise alone leaves inherited pipes open. The cleanup itself is useful and should stay.

File: sdk/packages/core/src/hooks/subprocess-runner.ts [hunk 7]
```diff
@@ -208,6 +283,71 @@ export async function runSubprocessEvent(
 		}, options.timeoutMs);
 	}
 
+	// Armed before the stdin write is awaited: a hook that never reads stdin
+	// can leave that write pending until it exits (seen on Windows), and an
+	// observation window that only starts afterwards would never censor
+	// anything — the sample would silently fall back to hooks that finished.
+	if (detached) {
+		if (options.onDetachedSettled) {
```
Comment (abeatrix) [inline sdk/packages/core/src/hooks/subprocess-runner.ts]:
```suggestion
	if (detached && options.onDetachedSettled) {
```

File: sdk/packages/agents/src/agent-runtime.ts (hunk 3 above)
Comment (mkondratek) [reply sdk/packages/agents/src/agent-runtime.ts]: done — one formatter, origin union. tool attrs only on the tool variant

Comment (mkondratek) [reply sdk/packages/agents/src/agent-runtime.ts]: nice finding

File: sdk/packages/agents/src/agent-runtime.ts (hunk 4 above)
Comment (mkondratek) [reply sdk/packages/agents/src/agent-runtime.ts]: good catch, deferral could drop it.  now it goes in right before the trailing tool call, so it always lands and the pair stays adjacent

File: sdk/packages/agents/src/agent-runtime.ts (hunk 5 above)
Comment (mkondratek) [reply sdk/packages/agents/src/agent-runtime.ts:889]: keep. moved back next to createMessage

File: sdk/packages/core/src/hooks/subprocess-runner.ts (hunk 6 above)
Comment (mkondratek) [reply sdk/packages/core/src/hooks/subprocess-runner.ts:277]: makes sense. done

**Review by abeatrix — APPROVED** [review]
